src/ui/swipe.ts

```diff
@@ -0,0 +1,164 @@
+import {DomWatcher} from '../dom/dom-watcher';
+import {EventCallback, EventManager, EventDispatcher} from './events';
+import * as dom from '../dom/dom';
+
+export enum SwipeEvent {
+  UP = 'swipeUp',
+  DOWN = 'swipeDown',
+  LEFT = 'swipeLeft',
+  RIGHT = 'swipeRight',
+}
+
+/**
+ * A basic class that allows you to watch directional swipe events on a given
+ * element.
+ *
+ *
+ * Usage:
+ * ```ts
+ * const swiper = new Swipe(myElement);
+ *
+ * swiper.on(Swipe.Events.UP, ()=> {
+ *   // swipe up.
+ * })
+ * swiper.on(Swipe.Events.LEFT, ()=> {
+ *   // swipe left
+ * })
+ *
+ * // Later
+ * swiper.dispose();
+ * ```
+ *
+ *
+ *
+ * Events are also fired on the element itself.
+ *
+ * ```ts
+ * const swiper = new Swipe(myElement);
+ * myElement.addEventListener(Swipe.Events.UP, ()=> {
+ *   // swipe up.
+ * })
+ * ```
+ */
+export class Swipe implements EventDispatcher {
+  private rootElement: HTMLElement;
+  private eventManager: EventManager;
+  private watcher: DomWatcher;
+  private x: Number;
+  private y: Number;
+  private isSwiping = false;
+  private allowSwipe = false;
+  private debounceDuration = 300;
+
+  constructor(rootElement: HTMLElement) {
+    this.rootElement = rootElement;
+    this.eventManager = new EventManager();
+    this.watcher = new DomWatcher();
+
+    const isTouchSupported = 'ontouchstart' in document.documentElement;
+
+    // Add listeners
+    if (isTouchSupported) {
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'touchstart',
+        eventOptions: {passive: true},
+        callback: this.onPointerDown.bind(this),
+      });
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'touchend',
+        eventOptions: {passive: true},
+        callback: this.onPointerUp.bind(this),
+      });
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'touchmove',
+        eventOptions: {passive: true},
+        callback: this.onPointerMove.bind(this),
+      });
+    } else {
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'mousedown',
+        eventOptions: {passive: true},
+        callback: this.onPointerDown.bind(this),
+      });
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'mouseup',
+        eventOptions: {passive: true},
+        callback: this.onPointerUp.bind(this),
+      });
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'mousemove',
+        eventOptions: {passive: true},
+        callback: this.onPointerMove.bind(this),
+      });
+    }
+  }
+
+  static Events = SwipeEvent;
+
+  private onPointerDown(e: TouchEvent) {
+    this.x = (e.touches && e.touches[0].clientX) || e['x'];
```

Comment:
Are you using `e['x']` because `x` isn't defined on `TouchEvent`? I wonder if you should declare `(e: TouchEvent | MouseEvent)` above in that case.

Reply:
Done - Had to check if it was a instance of to avoid typescript errors.